design-proposals/host-ip-pool.md

```diff
@@ -0,0 +1,503 @@
+changed to add more free IPs to the pool. Doing so involves that handlers will
+reconstruct `status.allocated-offsets` with the already reserved IPs at NNCE.
+
+### Drawbacks
+
+Updating the pool is distributed across nodes by the nmstate-handler pod, this
+means that N nodes will try to update `NodeIPPool` and retry on conflict. Is
+not expected the IPAM mechanism will run very often so this should not be a
+problem.
+
+Debugging the allocated offset can be a challange since it involve a convertion
+from the offset to the IP is pointing too depending on `NodeIPPool` spec.
+
+Fixing the "exhausted" issue updating the NodeIPPool involves a lot of logic to
+reconstruct the `allocated-offsets` that has to be carefully implemented and
+tested, what alternate would be to just delete `NodeIPPool` and create new one,
+so `NodeIPPool` is a read-only resource.
+
+The baremetal IPI Nmstate does an offline NetworkManager keyfile generation so
+there is no network `currentState` that nmpolicy can reference to, so using
+nmpolicy just to reference the ippools may be too overkill.
+
+If in the future nmstatectl is included at FCOS then the `networkPolicy` will
+be calculated online (but nmpolicyctl will have to be included at FCOS too).
+
+At non graceful node deletion the delete eventmay not be received at the
+controller so their IPs will not be deallocated, in that case manual
+intervention is needed.
+
+[2] https://github.com/coreos/fedora-coreos-tracker/issues/1175
+
+## Implementation History
+- Generic IPAM mechanism with in memory storage mechanism as either, standalone project, nmpolicy feature or kubernets-nmstate package.
+- IPAM storage mechanism for kubernetes-nmstate use case (day 2)
+- Expression "ippool.[name].allocate" feature either at kubernetes-nmstate or nmpolicy they have to call the created IPAM mechanism
+- Integrate IPAM, nmpolicy and kubernetes-nmstate for simplest use case, allocate an address for the first time.
+- Cover the deallocation scenario where the pool is unreferenced
+- Cover the scale up scenario
+- Cover the scale down scenario
+- Cover restart scenarios
+- Add `networkPolicy` field at baremetal API [openshift-installer API](https://github.com/openshift/installer/blob/master/pkg/types/baremetal/platform.go#L85)
+- Glue `networkPolicy` field with IPAM + nmpolicy at [image-customization-controller](https://github.com/openshift/image-customization-controller)
+- Test baremetal IPI
+- Add automatic garbage collector for stale IPs
+
+## Alternatives
+
+An alternative is to implement directly an IPAM feature at Nmstate with a
+plugin system for ippool retrieval and allocation storage so at day 2
+it will use the `NodeIPPool` and `NodeNetworkConfigurationEnactment` and
+at baremetal IPI It will read the ippool from the openshift installer config
+and store the allocations at the nodes file system, later on it will be
+converted into proper NodeIPPool and NNCEs using different storage plugin.
```

Comment:
I miss a comparison between the proposed solution and the alternative. 

Why did you choose one over the other ? What were its advantages over this one ?

Reply:
Well there are not much alternatives after looking around.